### PR TITLE
add device parameters processing

### PR DESCRIPTION
**Pull request by DeadByElpy**

### CODE REVIEW

File: src/CommandBinaryBuffer.ts [hunk 1]
```diff
@@ -177,6 +177,61 @@ interface IParameterOutputDataType {
...
+
+/**
+ * RX2 configuration.
+ * deviceParameters.RX_2_CONFIG = `18`.
+ *
+ * @see https://www.thethingsindustries.com/docs/reference/glossary/
+ */
+interface IParameterRx2Config {
+    /**
+     * The transmission speed or Data Rate of a LoRaWAN message, ranging from SF7 (highest Data Rate) to SF12 (lowest Data Rate).
```
Comment (DarkPark) [inline src/CommandBinaryBuffer.ts:215]:
```suggestion
     * The transmission speed or Data Rate of a LoRaWAN message, ranging from `SF7` (highest Data Rate) to `SF12` (lowest Data Rate).
```
?

File: src/CommandBinaryBuffer.ts [hunk 2]
```diff
@@ -216,6 +271,18 @@ interface IParameterAbsoluteDataStatus {
     status: number
 }
 
+/**
+ * Interval in seconds when device send EXTRA FRAME.
+ * deviceParameters.EXTRA_FRAME_INTERVAL = `28`.
+ */
+interface IParameterExtraFrameInterval {
+    /**
+     * If value = `0` EXTRA FRAME disabled.
```
Comment (DarkPark) [inline src/CommandBinaryBuffer.ts:280]:
```suggestion
     * If value is `0` EXTRA FRAME disabled.
```
?

as it's not code but sentence

File: src/CommandBinaryBuffer.ts [hunk 3]
```diff
@@ -405,6 +479,85 @@ class CommandBinaryBuffer extends BinaryBuffer {
...
+            byte2 = view.getUint8(2);
+        }
+
+        const value = (byte0 << 16) | (byte1 << 8) | byte2;
+
+        // setup sign if value negative
+        if ( byte0 & 0x80 ) {
+            return value - (1 << 24);
+        }
+
```
Comment (DarkPark) [inline src/CommandBinaryBuffer.ts:523]:
```suggestion
```

File: src/CommandBinaryBuffer.ts [hunk 4]
```diff
@@ -405,6 +479,85 @@ class CommandBinaryBuffer extends BinaryBuffer {
...
+
+        return value;
+    }
+
+    setUint12 ( value: number, isLittleEndian = this.isLittleEndian ): void {
+        console.log('setUint12', value);
```
Comment (DarkPark) [inline src/CommandBinaryBuffer.ts:529]:
```suggestion
```

File: src/constants/deviceParameters/rx2SpreadFactors.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+/**
```
Comment (DarkPark) [inline src/constants/deviceParameters/rx2SpreadFactors.ts:1]: do we really need this subdir?

if yes then deviceParameters.ts should be renamed to index.ts a moved here as well?

File: src/constants/deviceParameters/rx2SpreadFactors.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+ * Possible values of RX2 spread factor
```
Comment (DarkPark) [inline src/constants/deviceParameters/rx2SpreadFactors.ts:2]:
```suggestion
 * Possible values of RX2 spread factor
 *
 * @packageDocumentation
```

File: src/commands/downlink/SetParameter.ts [hunk 7]
```diff
@@ -39,6 +48,14 @@ const examples: TCommandExampleList = [
         },
         hex: {header: '03 02', body: '09 01'}
     },
+    {
+        name: 'set spread factor and frequency for RX_2 window',
```
Comment (DarkPark) [inline src/commands/downlink/SetParameter.ts:52]:
```suggestion
        name: 'set spread factor and frequency for RX2 window',
```
?

File: src/CommandBinaryBuffer.ts (hunk 1 above)
Comment (DeadByElpy) [reply src/CommandBinaryBuffer.ts:215]: fixed

File: src/CommandBinaryBuffer.ts (hunk 2 above)
Comment (DeadByElpy) [reply src/CommandBinaryBuffer.ts:280]: fixed

File: src/CommandBinaryBuffer.ts (hunk 3 above)
Comment (DeadByElpy) [reply src/CommandBinaryBuffer.ts:523]: fixed

File: src/CommandBinaryBuffer.ts (hunk 4 above)
Comment (DeadByElpy) [reply src/CommandBinaryBuffer.ts:529]: removed

File: src/constants/deviceParameters/rx2SpreadFactors.ts (hunk 5 above)
Comment (DeadByElpy) [reply src/constants/deviceParameters/rx2SpreadFactors.ts:1]: moved to root constants directory

File: src/commands/downlink/SetParameter.ts (hunk 7 above)
Comment (DeadByElpy) [reply src/commands/downlink/SetParameter.ts:52]: fixed

File: src/CommandBinaryBuffer.ts [hunk 8]
```diff
@@ -177,6 +177,61 @@ interface IParameterOutputDataType {
...
+/**
+ * RX2 configuration.
+ * deviceParameters.RX_2_CONFIG = `18`.
+ *
+ * @see https://www.thethingsindustries.com/docs/reference/glossary/
+ */
+interface IParameterRx2Config {
+    /**
+     * The transmission speed or Data Rate of a LoRaWAN message, ranging from `SF7` (highest Data Rate) to `SF12` (lowest Data Rate).
+     * Making the spreading factor 1 step lower (from SF10 to SF9) allows you to roughly send the same amount of data use half the time on air.
```
Comment (DarkPark) [inline src/CommandBinaryBuffer.ts:216]:
```suggestion
     * Making the spreading factor `1` step lower (from `SF10` to `SF9`) allows you to roughly send the same amount of data use half the time on air.
```

File: src/constants/deviceParameters/rx2SpreadFactors.ts (hunk 6 above)
Comment (DarkPark) [reply src/constants/deviceParameters/rx2SpreadFactors.ts]: are you sure?

Comment (DeadByElpy) [reply src/constants/deviceParameters/rx2SpreadFactors.ts]: now - added

File: src/CommandBinaryBuffer.ts (hunk 8 above)
Comment (DeadByElpy) [reply src/CommandBinaryBuffer.ts]: fixed

File: src/constants/deviceParameters.ts [hunk 9]
```diff
@@ -11,15 +11,24 @@ export const DAY_CHECKOUT_HOUR = 4;
...
 export const OUTPUT_DATA_TYPE = 5;
 
+/** type of priority data delivery, with confirmation or not */
+export const DELIVERY_TYPE_OF_PRIORITY_DATA = 8;
+
 /** device activation method */
 export const ACTIVATION_METHOD = 9;
 
+/** setup parameters for RX2 window config */
+export const RX_2_CONFIG = 18;
```
[the remark this answers is not in the document]
Comment (DeadByElpy) [reply src/constants/deviceParameters.ts]: fixed

**Review by DarkPark — APPROVED** [review]

**Pull request merged**